---
title: "Figure4"
author: "Raheel R Ahmad"
output: html_notebook
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)

rm(list = ls())

# Data Repo
knitr::opts_knit$set(root.dir = normalizePath("/home/jon/Documents/weng_lab/figure_repo")) 

```

```{r Figure4A}
library(tidyverse)
library(gtools)
library(glmnet)

# external scripts (in /scripts folder on GitHub repo)
source('./scripts-Project_Independent/matrix_functions.R')
source('./scripts-Project_Independent/regression_functions.R')
source('./scripts-Project_Independent/mixed_effect_regressionForest.R')
source('./scripts-Project_Independent/mixed_effect_regressionForest_ranger.R')
source('./scripts-Project_Independent/mixed_effect_elasticNet_glmnet_parallel.R')
source('./scripts-Project_Independent/machine_learning_functions.R')
library(ranger)
library(Metrics)
library(ggridges)

version <- '3.6.0'

# pull from /data/TCR/10X_Genomics/scRNAseq_P4_HUMAN_GEX/data/interim/MERF_ranger
rf.model <- readRDS('./3.5.5_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_30dim1500genes_fixed_OneGrp_logNorm.Rds')
enet.model <- readRDS("./3.5.5_MEEN_glmnet_integrated_OneGrp_logNorm_parallel_geneNameFix.Rds")
df <- readRDS('./3.5.5_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_30dim1500genes_fixed_DATA_OneGrp_logNorm.Rds')

# ENET:
fit.enet <- enet.model$fit.enet
currInd <- match(fit.enet$lambda.min,fit.enet$glmnet.fit$lambda)
predictions <- as.numeric(fit.enet$fit.preval[,currInd])
bi <- as.numeric(enet.model$`bi[[r]]`)
bi_data <- tibble(bi = bi, ID = unique(df$cluster.id))

# MERF (For Mixed Effect Random Forest Graphs):
# predictions <- rf.model$fit.ranger$predictions
# bi <- rf.model$`bi[[r]]`
# bi_data <- tibble(bi = bi, ID = unique(df$cluster.id))

predict_df <- tibble(Age = df$Y, ID = df$cluster.id,  prediction = predictions) %>% inner_join(bi_data, by = 'ID')
#group = factor(ID)
#predict_df <- predict_df %>% group_by(ID,Age) %>% filter(prediction < quantile(prediction, 0.99),  ## this appears to be breaking it
#                                                       prediction > quantile(prediction, 0.01))

predict_df <- predict_df %>% group_by(ID,Age)
predict_df <- predict_df %>% mutate(facet = ifelse(ID == 'cross-sectional', 'cross-sectional', 'longitudinal'))
predict_df$facet <- factor(predict_df$facet, levels = c('longitudinal', 'cross-sectional'))
predict_df$group <- paste(predict_df$Age, predict_df$facet, sep='_')
#write.csv(predict_df, '/data/TCR/10X_Genomics/scRNAseq_P4_HUMAN_GEX/data/interim/MERF_ranger/3.5.4_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_TreatAllCrossSectional_predictions.Rds')
# for MERF prediction: y = prediction + bi

m.p<-ggplot(predict_df, aes(x=Age, y = prediction + bi)) +
  theme_minimal() + theme_bw(base_size = 7)  + 
  theme(aspect.ratio = 1, legend.position = 'none', axis.title.x = element_blank(), axis.title.y = element_blank()) +
  geom_abline(intercept = 0, slope = 1, color = 'grey40', linetype = 2) +
  #geom_violin(scale = 'width', aes(group=factor(Age)), fill = 'grey60') +
  geom_boxplot(width = 0.7, aes(group=group, fill = facet), lwd = 0.25, color = 'grey30', outlier.alpha = 0) +
  #geom_boxplot(width = 0.5, aes(group=factor(Age))) +
  #geom_jitter(width = 1) +
  geom_smooth(data = predict_df[predict_df$facet == 'longitudinal',], method = 'lm', aes(group = ID, color = 'darkred',alpha = 5), size = 0.25, se = F) +
  #stat_smooth(geom='line', alpha=0.5, se=FALSE) +
  geom_smooth(method = 'lm', size = 0.65, color = 'darkblue') + 
  coord_cartesian(xlim = c(-10,96), ylim = c(-10,95)) +
  scale_x_continuous(breaks = c(0,20,40,60,80,100)) +
  scale_y_continuous(breaks = c(0,20,40,60,80,100))
  #scale_fill_manual(values = c('goldenrod3','forestgreen'))
#facet_wrap(ID~., nrow = 1, scales = 'free_x')
print("Figure 4A")
ggsave("fig4a.png")
print(m.p)
write.csv(predict_df, "fig4a_predict_df.csv")

## DENSITY PLOTS
for (group_id in unique(predict_df$group)){
  predict_sub <- filter(predict_df, group==group_id)
  density_plot <- ggplot(predict_sub, aes(x=prediction)) + 
    geom_density() +
    geom_vline(aes(xintercept=unique(Age)), 
               color="blue", linetype='dashed', size=1)
  print(density_plot)
  dir.create('density_plots', showWarnings = F)
  ggsave(paste0("density_plots/density_", group_id, ".png"))
  #break
}

```


```{r Figure4B_1}
### part 1 predicts on a validation portion of training data source
library(tidyverse)
library(gtools)
library(glmnet)

# external scripts (in /scripts folder on GitHub repo)
source('./scripts-Project_Independent/matrix_functions.R')
source('./scripts-Project_Independent/regression_functions.R')
source('./scripts-Project_Independent/mixed_effect_regressionForest.R')
source('./scripts-Project_Independent/mixed_effect_regressionForest_ranger.R')
source('./scripts-Project_Independent/mixed_effect_elasticNet_glmnet_parallel.R')
source('./scripts-Project_Independent/machine_learning_functions.R')

library(ranger)
library(Metrics)
library(ggridges)

# pull from /data/TCR/10X_Genomics/scRNAseq_P4_HUMAN_GEX/data/interim/MERF_ranger
rf.model <- readRDS('./3.5.5_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_30dim1500genes_fixed_OneGrp_logNorm.Rds')
enet.model <- readRDS("./3.5.5_MEEN_glmnet_integrated_OneGrp_logNorm_parallel_geneNameFix.Rds")
df <- readRDS('./3.5.5_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_30dim1500genes_fixed_DATA_OneGrp_logNorm.Rds')

# for ENET:
fit.enet <- enet.model$fit.enet
currInd <- match(fit.enet$lambda.min,fit.enet$glmnet.fit$lambda)
predictions <- as.numeric(fit.enet$fit.preval[,currInd])
bi <- as.numeric(enet.model$`bi[[r]]`)
bi_data <- tibble(bi = bi, ID = unique(df$cluster.id))

predictions_meta <- read_csv('./3.5.5_MEEN_glmnet_integrated_META_OneGrp_logNorm_parallel.csv', guess_max = 100000)
all.validation <- fit.enet$fit.preval[,67] # get the validation predictions for all cells

validation.df <- data.frame(prediction=all.validation,
  Age=predictions_meta$Age) %>%
  filter(Age>5&(Age<30|Age>70)) %>%
  mutate(Age_Group=ifelse(Age>40,"Old","Young"))
validation.df$Age_Group <- factor(validation.df$Age_Group, levels = c('Young','Old'))

m.p<-ggplot(validation.df, aes(x=Age_Group, y=prediction)) +
  geom_violin(aes(fill = Age_Group)) +
  scale_fill_manual(values = c('pink', 'darkred')) +
  theme_minimal() + theme_bw(base_size = 6) +
  geom_boxplot(width=0.1) +
  coord_cartesian(ylim = c(3,97)) +
  scale_y_continuous(breaks = c(0,20,40,60,80,100))

print("Figure 4B_1")
ggsave("fig_4b_1.png")
write.csv(validation.df, "fig4b_validation_df.csv")
print(m.p)

# MERF Graphs:
predictions <- rf.model$fit.ranger$predictions
bi <- rf.model$`bi[[r]]`
bi_data <- tibble(bi = bi, ID = unique(df$cluster.id))

predict_df <- tibble(Age = df$Y, ID = df$cluster.id,  prediction = predictions) %>% inner_join(bi_data, by = 'ID')

predict_df <- predict_df %>% group_by(ID,Age)
predict_df <- predict_df %>% mutate(facet = ifelse(ID == 'cross-sectional', 'cross-sectional', 'longitudinal'))
predict_df$facet <- factor(predict_df$facet, levels = c('longitudinal', 'cross-sectional'))
predict_df$group <- paste(predict_df$Age, predict_df$facet, sep='_')
# for MERF prediction: y = prediction + bi

predict_df <- predict_df %>% filter(Age > 5 & Age <= 30 | Age >= 70)
predict_df <- predict_df %>% mutate(Age_Group = ifelse(Age > 40, 'Old', 'Young'))
predict_df$Age_Group <- factor(predict_df$Age_Group, levels = c('Young','Old'))
m.p<-ggplot(predict_df, aes(x=Age_Group, y=prediction + bi)) +
  geom_violin(aes(fill = Age_Group)) +
  scale_fill_manual(values = c('pink', 'darkred')) +
  theme_minimal() + theme_bw() +
  geom_boxplot(width=0.1) +
  coord_cartesian(ylim = c(3,97)) +
  scale_y_continuous(breaks = c(0,20,40,60,80,100))
print("Supplemental Figure4D")
ggsave("supp_fig_4d.png")
write.csv(validation.df, "fig4b_predict_df.csv")
print(m.p)
```


```{r Figure4B_2}
# on external data (Fig4B_2):
# Restart R for package compatibility!!!!! (restarting unloads all packages)

library(tidyverse)
library(Seurat)
library(ranger)
library(psych)
library(tensorflow)
library(keras)
library(glmnet)
require(methods)

suppressMessages({
test_list <- read_csv('./4.4.0_test_list.csv')
})
test_list$sample <- c(1:15)
rf_data <- readRDS('./3.5.5_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_30dim1500genes_fixed_DATA_OneGrp_logNorm.Rds')
final <- readRDS('./4.4.0_TestMatrix_geneNameFix.Rds')
labels <- abs(parse_number(rownames(final)))

# RF Model:
RF.model <- readRDS('./3.5.5_MERF_ranger_500tree_3000mtry_100ndsize_IntegratedData_30dim1500genes_fixed_OneGrp_logNorm.Rds')
RF.fit <- RF.model$fit.ranger
# ENET Model:
enet.model <- readRDS("./3.5.5_MEEN_glmnet_integrated_OneGrp_logNorm_parallel.Rds")
fit.enet <- enet.model$fit.enet

# predict on raw log2 data:
#final_tbl <- tibble(final)
#final_mat <- data.matrix(final)
#final_dgc <- as(final_mat, "dgCMatrix")
#index <- convert_genes(colnames(final_tbl), rownames(data))
# write.csv(final_tbl, './4.4.0_TestMatrix_geneNameFix.csv')

predictions_log2 <- predict(fit.enet, newx = final, s="lambda.min")
pred_tbl <- tibble(sample=labels, prediction=predictions_log2[,1])
pred_tbl <- pred_tbl %>% inner_join(test_list)
pred_tbl$Age_Group <- pred_tbl$Age
pred_tbl$Age_Group <- factor(pred_tbl$Age_Group, levels = c('Young','Old'))
m.p<-ggplot(pred_tbl, aes(x=factor(Age_Group), y=prediction)) +
  geom_violin(aes(fill = Age_Group))+
  scale_fill_manual(values = c('pink', 'darkred')) +
  theme_minimal() + theme_bw(base_size = 6) +
  geom_boxplot(width=0.1) +
  coord_cartesian(ylim = c(3,97)) +
  scale_y_continuous(breaks = c(0,20,40,60,80,100))


print("Figure 4B_2")
ggsave("Figure_4B_2.png")
print(m.p)
write.csv(pred_tbl, "fig4b_2_pred_tbl.csv")

```
```{r Compile USCMD Output}
library(tidyverse)
library(argparse)
library(MASS)

compile_steps <- function(donor_csv, outpath_11, outpath_9) {
  donor_df <- read.csv(donor_csv)
  # print("donor_df")
  # print(donor_df)

  pipeline_dir <- donor_df[1,4]
  sample_name <- donor_df[1, 1]
  print(sample_name)
  step11_path <- file.path(pipeline_dir, "step11", "mutations.csv")
  tot.df.11 <- read.csv(step11_path, header=F)
  tot.df.11$sample <- sample_name

  step9_path <- file.path(pipeline_dir, "step9", "filtered_ScoredMutations.csv")
  tot.df.9 <- read.csv(step9_path, header=T)
  tot.df.9$sample <- sample_name

  for (i in 2:nrow(donor_df)) {
    pipeline_dir <- donor_df[i,4]
    sample_name <- donor_df[i, 1]
    print(sample_name)
    step11_path <- file.path(pipeline_dir, "step11", "mutations.csv")
    print(step11_path)
    step11_csv <- read.csv(step11_path, header=F)
    step11_csv$sample <- sample_name
    tot.df.11 <- rbind(tot.df.11, step11_csv)

    step9_path <- file.path(pipeline_dir, "step9", "filtered_ScoredMutations.csv")
    step9_csv <- read.csv(step9_path, header=T)
    step9_csv$sample <- sample_name
    tot.df.9 <- rbind(tot.df.9, step9_csv)
  }
  colnames(tot.df.11) <- c("Cell", "exome_reads", "total_reads", 
    "umis", "exome_cov", "total_cov", "sample")
  ######## load in MD so that batches get separated into visits  
  meta <- read.csv("./clustering_metadata_logNorm.csv")

  meta <- filter(meta, study=="longi") %>% 
    #dplyr::select(barcode, Person, Code, visit, Age) %>%
    mutate(bc=substr(barcode, 1, 16), Code.visit=paste0(Code, ".", visit))
  tot.df.11$barcode <- str_extract(tot.df.11$Cell, "[ATCG]{16}-[:digit:]")
  tot.df.11 <- mutate(tot.df.11, bc=substr(barcode,1,16))%>%
    mutate(Code=substr(sample, 1, 2))

  tot.df.11 <- left_join(tot.df.11, meta, by=c("Code", "bc")) 
  write.csv(tot.df.11, outpath_11, row.names=F)

  tot.df.9 <- tot.df.9 %>%
    mutate(Code=substr(donor, 1, 2), bc=substr(bc, 1, 16))

  ######## load in MD so that batches get separated into visits  
  meta <- read_csv('./clustering_metadata_logNorm.csv', guess_max = 100000)
  meta <- filter(meta, study=="longi") %>% 
    dplyr::select(barcode, Person, Code, visit, Age) %>%
    mutate(bc=substr(barcode, 1, 16), Code.visit=paste0(Code, ".", visit))

  tot.df.9 <- left_join(tot.df.9, dplyr::select(meta, Code, bc, Code.visit, Person), by=c("Code", "bc"))
  write.csv(tot.df.9, outpath_9)
}

compile_steps_cross <- function(donor_csv, outpath_11, outpath_9) {
  donor_df <- read.csv(donor_csv)
  pipeline_dir <- donor_df[1,4]
  sample_name <- donor_df[1, 1]
  print(sample_name)
  step11_path <- file.path(pipeline_dir, "step11", "mutations.csv")
  tot.df.11 <- read.csv(step11_path, header=F)
  tot.df.11$sample <- sample_name

  step9_path <- file.path(pipeline_dir, "step9", "filtered_ScoredMutations.csv")
  tot.df.9 <- read.csv(step9_path, header=T)
  tot.df.9$sample <- sample_name

  for (i in 2:nrow(donor_df)) {
    pipeline_dir <- donor_df[i,4]
    sample_name <- donor_df[i, 1]
    print(sample_name)
    step11_path <- file.path(pipeline_dir, "step11", "mutations.csv")
    print(step11_path)
    step11_csv <- read.csv(step11_path, header=F)
    step11_csv$sample <- sample_name
    tot.df.11 <- rbind(tot.df.11, step11_csv)

    step9_path <- file.path(pipeline_dir, "step9", "filtered_ScoredMutations.csv")
    step9_csv <- read.csv(step9_path, header=T)
    step9_csv$sample <- sample_name
    tot.df.9 <- rbind(tot.df.9, step9_csv)
  }
  colnames(tot.df.11) <- c("Cell", "exome_reads", "total_reads", 
    "umis", "exome_cov", "total_cov", "sample") 

  tot.df.11$barcode <- str_extract(tot.df.11$Cell, "[ATCG]{16}-[:digit:]")
  tot.df.11 <- mutate(tot.df.11, bc=substr(barcode,1,16))%>%
    mutate(Code=substr(sample, 1, 2))

  write.csv(tot.df.11, outpath_11, row.names=F)

  write.csv(tot.df.9, outpath_9)
}

donor_list_longi <- "longi_donors.csv"
donor_list_cross <- "cross_donors.csv"

compile_steps(donor_list_longi,
  "compiled_11_longi.csv",
  "compiled_9_longi.csv")
compile_steps_cross(donor_list_cross,
  "compiled_11_cross.csv",
  "compiled_9_cross.csv")
```


```{r Figure4C/D/E_Analysis}
# 7.0.0_ageChange
rm(list=ls())
library(tidyverse)
library(mpath)
library(glmnet)
library(glmnetUtils)
library(MASS)

# Load Data:
genome_basepair_size <- 3099750718

# mutations:
suppressMessages({ # suppress the csv loading messages
# load step 9 output
mutations_cross <- read_csv('./compiled_9_cross.csv', guess_max = 100000) %>%
   filter(is.na(umi_fraction_filter) == F) %>% dplyr::select(Chr, POS, REF, ALT, GENE, donor, bc)
mutations_longi <- read_csv('./compiled_9_longi.csv', guess_max = 100000) %>%
   filter(is.na(umi_fraction_filter) == F) %>% dplyr::select(Chr, POS, REF, ALT, GENE, donor, bc)

colnames(mutations_cross)[c(6)] <- c('BL_ID')
colnames(mutations_longi)[c(6)] <- c('Person')
mutations_cross$bc <- substr(mutations_cross$bc, 1,16)
mutations_longi$bc <- substr(mutations_longi$bc, 1,16)
  
  # metadata:
meta <- read_csv('./clustering_metadata_logNorm.csv', guess_max = 100000)
meta$bc <- paste(substr(meta$barcode, 1,16), '-1', sep='')
pheno <- read_csv('./pheno_assgnmnt_logNorm.csv')
meta <- meta %>% inner_join(pheno) %>% filter(cell != "NK")

# remove cells seen in multiple libraries:
# Also checked on line 319 and 429
doubled_up_cells <- group_by(meta, BL_ID, bc, visit) %>% 
  tally() %>% filter(n>1) %>% left_join(meta, by=c("BL_ID", "bc", "visit"))
write.csv(doubled_up_cells, "./doubled_up_cells.csv")

meta <- meta %>% anti_join(doubled_up_cells, by = c("BL_ID", "bc", "visit"))
meta$bc <- str_sub(meta$bc, 1, 16)
meta_basic_cross <- meta[,c(6,17)] %>% filter(str_detect(BL_ID, 'BL'))
meta_basic_longi <- meta %>% filter(!str_detect(BL_ID, 'BL'))
meta_basic_longi <- meta_basic_longi[,c(3,17)]
  
# predictions:
predictions <- read_csv('./4.3.0_MERF_MEEN_residuals_predictions.csv', guess_max = 100000) %>%
  filter(model == 'MEEN')
predictions_meta <- read_csv('./3.5.5_MEEN_glmnet_integrated_META_OneGrp_logNorm_parallel.csv', guess_max = 100000)
predictions <- cbind(predictions[,2:5],predictions_meta$barcode) # <-- GET ERROR HERE

  # read, umi, coverage information (step11):
read_cross <- read.csv('./compiled_11_cross.csv', header=TRUE)
read_longi <- read.csv('./compiled_11_longi.csv', header=TRUE)
doubled_up_reads <- read_longi %>% group_by(Person, bc, visit) %>% summarize(num = n()) %>% filter(num > 1)
read_longi <- read_longi %>% anti_join(doubled_up_reads, by = c("Person", "bc", "visit"))

read_cross <- dplyr::rename(read_cross, c("BL_ID"="sample" , "reads_exome" = "exome_reads", "reads"="total_reads", "umi" = "umis", "coverage_exome" = "exome_cov", "coverage"="total_cov"))
read_cross$bc <- str_extract(read_cross$bc, regex("[ACGT]+.*"))
read_cross <- distinct(read_cross, BL_ID, bc, reads_exome, reads, umi, coverage_exome, coverage)

read_longi <- rename(read_longi, c("reads_exome" = "exome_reads", "reads"="total_reads", "umi" = "umis", "coverage_exome" = "exome_cov", "coverage"="total_cov"))
read_longi$bc <- str_extract(read_longi$bc, regex("[ACGT]+.*"))
read_longi <- distinct(read_longi, Person, bc, reads_exome, reads, umi, coverage_exome, coverage)

exome_cross <- read_csv('./donor_summary_cross.csv', guess_max = 100000)
exome_longi <- read_csv('./donor_summary_longi.csv', guess_max = 100000)

# Combine Studies:
mut_cross <- mutations_cross %>% group_by(BL_ID,bc) %>% summarize(num_mut = n())
mut_0 <- meta_basic_cross %>% left_join(mut_cross, by = c('BL_ID','bc')) 
mut_0$bc <- str_sub(mut_0$bc, 1, 16)
mut_0$num_mut[is.na(mut_0$num_mut)] <- 0
mut_cross_comb <- mut_0 %>% inner_join(read_cross, by = c('BL_ID','bc')) %>%
  inner_join(exome_cross, by = c('BL_ID')) %>%
  inner_join(meta, by = c('BL_ID', 'bc'))
  
mut_longi <- mutations_longi  %>% group_by(Person,bc) %>% summarize(num_mut = n()) 
mut_0 <- meta_basic_longi %>% left_join(mut_longi, by = c('Person','bc')) %>% mutate(Code = substr(Person, 1, 2))
mut_0$bc <- str_sub(mut_0$bc, 1, 16)
mut_0$num_mut[is.na(mut_0$num_mut)] <- 0
mut_longi_comb <- mut_0  %>% inner_join(read_longi, by = c('Person','bc')) %>%
  inner_join(exome_longi, by = c('Code')) %>%
  inner_join(meta, by = c('Code','Person', 'bc')) %>%
  anti_join(doubled_up_cells, by = c('Person','bc','visit','Code'))

mut_all <- bind_rows(mut_cross_comb, mut_longi_comb)
write.csv(mut_all, './7.0.0_Mutations_number_new.csv', row.names = F)
mut_all <- read_csv('./7.0.0_Mutations_number_new.csv', guess_max = 100000)

# replace coverage and exome coverage with corrected data at 10X depth:
new_coverage <- read_csv('./201127_alignment_bam_coverages_Jeffrey.csv')
colnames(new_coverage)[1] <- "Code"
})
mut_all <- mut_all %>% dplyr::select(-ex_coverage, -ex_coverage_exome) %>% inner_join(new_coverage, by = 'Code')

# Diagnostic Plots:
# Perform adjustments on log2 transformed data for all metrics
mut_all$log2_mut <- 0
mut_all$norm_mut <- 0
mut_all$scale_mut <- 0
adj_mut <- mut_all %>% filter(num_mut > 0)
kept_mut <- mut_all %>% filter(num_mut == 0)

# Adjust values for mutation number
adj_mut$log2_mut <- log2(adj_mut$num_mut+1)
  
# Plot these three: umi, coverage, ex_coverage
# UMI:
m.p <- ggplot(adj_mut, aes(x=umi, y=num_mut)) +
  theme_minimal() + theme_bw(base_size = 7) + theme(aspect.ratio = 1, legend.position = 'none') +
  geom_point(aes(color = study), size = 0.1) +
  geom_smooth(method="glm.nb", se=TRUE, color = 'black') +
  coord_cartesian(ylim = c(0, 35))
print(m.p)
ggsave('./Sup7A_7.0.0_diagnostic_NB_RawRm0_UmiVsMutations.png', height = 45, width = 45, units = 'mm', dpi = 300)

#png('./Sup7A_7.0.0_diagnostic_NB_RawRm0_UmiVsMutations.png', height = 45, width = 45, units = 'mm', res = 300)
#print(m.p)
#dev.off()

# Single Cell Coverage
m.p <- ggplot(adj_mut, aes(x=coverage, y=num_mut)) +
  theme_minimal() + theme_bw(base_size = 7) + theme(aspect.ratio = 1, legend.position = 'none') +
  geom_point(aes(color = study), size = 0.1) +
  geom_smooth(method="glm.nb", se=TRUE, color = 'black') +
  coord_cartesian(ylim = c(0, 35))

print(m.p)
ggsave('./Sup7B_7.0.0_diagnostic_NB_RawRm0_coverageVsMutations.png', height = 45, width = 45, units = 'mm', dpi = 300)

#png('./Sup7B_7.0.0_diagnostic_NB_RawRm0_coverageVsMutations.png', height = 45, width = 45, units = 'mm', res = 300)
#print(m.p)
#dev.off()

# Exome Coverage
m.p <- ggplot(adj_mut, aes(x=ex_coverage, y=num_mut)) +
  theme_minimal() + theme_bw(base_size = 7) + theme(aspect.ratio = 1, legend.position = 'none') +
  geom_point(aes(color = study), size = 0.1) +
  geom_smooth(method="glm.nb", se=TRUE, color = 'black') +
  coord_cartesian(ylim = c(0, 35))

print(m.p)
ggsave('./Sup7C_7.0.0_diagnostic_NB_RawRm0_ExomecoverageVsMutations.png', height = 45, width = 45, units = 'mm', dpi = 300)

#png('./Sup7C_7.0.0_diagnostic_NB_RawRm0_ExomecoverageVsMutations.png', height = 45, width = 45, units = 'mm', res = 300)
#print(m.p)
#dev.off()
summary(glm.nb(num_mut ~ coverage, data=adj_mut))
# Negative Binomial regression:
print("380")
print(dim(adj_mut))
adj_mut <- adj_mut[is.na(adj_mut$umi)==FALSE,]
print(dim(adj_mut))

suppressWarnings({
nbGLM <- glm.nb(log2_mut ~ umi + coverage + ex_coverage, data=adj_mut)
})

adj_mut$norm_mut = (nbGLM$residuals + nbGLM[["coefficients"]][["(Intercept)"]])

#png('./7.0.0_diagnostic_normMutations_histogram.png', height = 100, width = 100, units = 'mm', res = 300)
#hist(adj_mut$norm_mut)
#dev.off()
# Normalize to 0 to 1:
adj_quantile <- quantile(adj_mut$norm_mut, probs = c(0.001,0.999))
adj_mut[adj_mut$norm_mut < min(adj_quantile),]$norm_mut <- min(adj_quantile)
adj_mut[adj_mut$norm_mut > max(adj_quantile),]$norm_mut <- max(adj_quantile)
value <- adj_mut$norm_mut
adj_mut$scale_mut <- (value - min(value)) / (max(value) - min(value)) * max(adj_mut$num_mut) 

library(plotly)
library(plot3D)
library(RColorBrewer)

# Sup fig 7D
# Save from notebook output in conxole bwlow chunk:
# Right clich picture --> 'save as image'
print(scatter3D(z=adj_mut$coverage, y=adj_mut$umi, x=adj_mut$ex_coverage, 
        type="scatter3d", bty = "g", theta = 35, phi = 25,
        colvar=adj_mut$num_mut, size = 0.01, col = brewer.pal(10, 'RdBu')[c(1:2,7:9,10,10,10)]))

# Sup fig 7E
# Save from notebook output in conxole bwlow chunk:
# Right clich picture --> 'save as image'
print(scatter3D(z=adj_mut$coverage, y=adj_mut$umi, x=adj_mut$ex_coverage, 
        type="scatter3d", bty = "g", theta = 35, phi = 25,
        colvar=adj_mut$scale_mut, size = 0.01, col = brewer.pal(10, 'RdBu')[c(1:2,7:9,10,10,10)]))
suppressWarnings({
summary(glm.nb(data = adj_mut, scale_mut ~ umi + coverage + ex_coverage))
})

mut_all <- rbind(adj_mut, kept_mut)
mut_all <- mut_all %>% anti_join(doubled_up_cells, by = c('BL_ID','bc','visit'))
mut_all_no_NK <- filter(mut_all, subset != "NK")

# log2 transform to better approximate a normal distribution:
write.csv(mut_all, './7.0.0_Mutations_normalized_lmAdjusted_rm0.csv', row.names = F)
write.csv(mut_all_no_NK, './7.0.0_Mutations_normalized_lmAdjusted_rm0_no_NK.csv', row.names = F)

```

```{r Figure4C/D/E}
# 7.1.0_ageChange_2
rm(list=ls())
library(lme4)
library(broom.mixed)
library(lmerTest)
library(tidyverse)
library(MuMIn)

# Load Mutations Data:

mut_all <- read_csv('./7.0.0_Mutations_normalized_lmAdjusted_rm0.csv', guess_max = 100000) %>% 
  filter(cell == 'CD8')

predictions <- read_csv('./7.0.0_Mutations_predictions.csv')
mut_all <- mut_all %>% inner_join(predictions, by = 'barcode')
# Show Relationship to Age:

colors <- tibble(subset = c('N','CM','EM','RA'), color = c('forestgreen','dodgerblue3','orangered','purple3'))

# Filter unnecessary cells:
mut_all <- mut_all %>% filter(Age > 5, subset != 'EFF')
mut_all$Group <- substr(mut_all$Code, 1, 2)
mut_all$prediction_add <- mut_all$prediction + mut_all$bi

# Age vs Predicted Age by subsets:

mut_all$subset <- factor(mut_all$subset, levels = colors$subset)
m.p<-ggplot(mut_all, aes(x=Age,y=prediction_add)) +
  theme_minimal() + theme_bw(base_size=7) + theme(aspect.ratio = 1) +
  geom_smooth(method = 'lm', aes(group = subset, color = subset), se = F) +
  scale_color_manual(values = colors$color) +
  coord_cartesian(xlim = c(20,90), ylim = c(20,90)) +
  scale_x_continuous(breaks = seq(20,90,10)) +
  scale_y_continuous(breaks = seq(20,90,10))
ggsave("fig1d.png")
print(m.p)
# ggsave('./figures/7.1.0_AgeVsPredictedAge_subset.png', height = 75, width=95, units='mm', dpi =600)  
# mixed effect line:
final <- tibble()
for (i in unique(mut_all$subset)) {
  mut_sub <- mut_all[mut_all$subset == i,]
  model <- lmer(prediction_add~Age + (1|Group), REML = T, mut_sub) # This is the actual linear model generated, change this to preferred model if necessary
  #summary(model)
  sum_test <- step(model,reduce.fixed = FALSE, reduce.random = FALSE)
  summary <- broom.mixed::tidy(model)
  summary2 <- broom.mixed::augment(model)
  age_p <- sum_test[["fixed"]][["Pr(>F)"]][[1]]
  co <- as.numeric(summary[2,4])
  intercept <- as.numeric(summary[1,4])
  add <- tibble(subset = i, age_p = age_p, co = co, intercept = intercept)
  final <- bind_rows(final, add)
}
write.csv(final, './7.1.0_Mutations_AgeVsPredictedAge_subset_ME.csv', row.names = F)


# Plot positive cells for their Expression Across Age:
# plot log2 and raw data
# Age and Predicted Age

pos_mut <- mut_all %>% filter(num_mut > 0)

# generate line:
final <- tibble()
for (i in unique(mut_all$subset)) {
  mut_sub <- pos_mut[pos_mut$subset == i,]
  model <- lm(scale_mut~prediction_add, mut_sub)
  #summary(model)
  cor <- cor(mut_sub$scale_mut, mut_sub$prediction_add)
  sum_test <- summary(model)
  age_p <- sum_test[["coefficients"]][2,4]
  co <- sum_test[["coefficients"]][2,1]
  intercept <- sum_test[["coefficients"]][1,1]
  add <- tibble(subset = i, age_p = age_p, co = co, intercept = intercept, cor = cor)
  final <- bind_rows(final, add)
}
write.csv(final, './7.1.0_log2ScalePosNumMutationsVsPredictedAge_subset_LM_RawRm0NB.csv', row.names = F)

final$subset <- factor(final$subset, levels = colors$subset)
m.p<-ggplot(pos_mut, aes(x=prediction_add, y=scale_mut)) +
  theme_minimal() + theme_bw(base_size = 7) + theme(aspect.ratio = 1) +
  geom_point(alpha = 0) +
  #geom_boxplot(aes(fill = factor(Age))) +
  geom_abline(data = final, aes(slope = co, intercept = intercept, color = subset)) +
  coord_cartesian(xlim = c(20,90), ylim = c(12,18)) +
  scale_x_continuous(breaks = seq(20,90,10)) +
  scale_color_manual(values = colors$color) +
  scale_y_continuous(breaks = seq(7,20,1))
ggsave("fig4e.png")
print(m.p)
ggsave('./7.1.0_PosNumMutationsVsPredictedAge_subset_ME_Log2Rm0.png', height = 75, width=95, units='mm', dpi =600)
write.csv(pos_mut, "fig4e_pos_mut.csv")
```


```{r Figure4E}
## get HIV age predictions

library(Matrix)

library(tidyverse)
library(Seurat)
library(ranger)
library(psych)
library(tensorflow)
library(keras)
library(glmnet)

require(methods)
# BiocManager::install("org.Hs.eg.db")

add_dummy_genes_to_matrix <- function(data_matrix, gene_list_original) {
  
  # identify missing genes:
  pbmc_genes <- rownames(data_matrix)
  genes_dummy <- gene_list_original[!(gene_list_original %in% pbmc_genes)]
  genes_absent <- pbmc_genes[!(pbmc_genes %in% gene_list_original)]

  print(length(genes_dummy))
  print(length(genes_absent))
  
  # identify all possible names of genes that are 'missing'
  library(org.Hs.eg.db)
  convert <- AnnotationDbi::select(org.Hs.eg.db, keys = genes_absent, columns="SYMBOL",keytype="ALIAS")
  
  # extract the names of the genes that match
  converted_genes <- convert[convert$SYMBOL %in% genes_dummy,]
  colnames(converted_genes)[1] <- 'gene'
  converted_genes <- converted_genes %>% group_by(SYMBOL) %>% sample_n(1)
  converted_genes <- converted_genes[converted_genes$gene %in% pbmc_genes,]
  converted_genes <- converted_genes[converted_genes$SYMBOL %in% gene_list_original,]
  pbmc_df <- tibble(gene=pbmc_genes) %>% left_join(converted_genes, by = 'gene')
  pbmc_df <- pbmc_df %>% group_by(gene) %>% sample_n(1) %>% ungroup()
  pbmc_df2 <- tibble(gene=pbmc_genes) %>% left_join(pbmc_df, by = 'gene')
  pbmc_df <- pbmc_df2 %>% mutate(gene = ifelse(is.na(SYMBOL), gene, SYMBOL))
  
  newrownames <- pbmc_df$gene
  genes_dummy <- gene_list_original[!(gene_list_original %in% newrownames)]
  
  
  empty_mat <- matrix(rep(0, length(genes_dummy) * ncol(data_matrix)), ncol = ncol(data_matrix))
  rownames(empty_mat) <- genes_dummy
  
  
  rownames(data_matrix) <- newrownames
  test_data <- rbind(data_matrix, empty_mat)
  test_data <- test_data[gene_list_original,]
  
  return(test_data)
  
}

seu_obj <- readRDS("./CD8_cells.Rds")
all.dat <- seu_obj[['RNA']]@data
all.dat <- all.dat[(rownames(all.dat) != "Sex"),]
all.dat <- rbind(all.dat, Sex=rep(0,ncol(all.dat))) # 0 for female, 1 for male.

######
enet.model <- readRDS('./3.5.5_MEEN_glmnet_integrated_OneGrp_logNorm_parallel_geneNameFix.Rds')
fit.enet <- enet.model$fit.enet

model.features <- fit.enet[["glmnet.fit"]][["beta"]]@Dimnames[[1]]

print(dim(all.dat))
print(class(all.dat))
all.dat <- t(add_dummy_genes_to_matrix(all.dat, model.features))
saveRDS(all.dat, './test_data_figure4E.Rds')
print(dim(all.dat))
print(class(all.dat))

all.dat.mat <- as.matrix(all.dat)

prediction <- predict(fit.enet, newx = all.dat.mat, s="lambda.min")
saveRDS(prediction, "age_predictions_CD8.Rds")

## load in and plot the predictions
library(tidyverse)
library(Seurat)
library(Matrix)
library(rstatix)
library(ggpubr)

seu_metadata <- readRDS(file = "./CD8_cells_metadata.Rds")
metadata <- as_tibble(read.table(file = './alexandria_structured_metadata.txt', 
                       sep = '\t', header = TRUE))
age_predictions <- readRDS("./age_predictions_CD8.Rds")
clust <- seu_metadata$seurat_clusters

wanted_meta <- dplyr::select(metadata, TimePoint, organism_age, donor_id)
row.names(wanted_meta) <- metadata$NAME
df <- merge(wanted_meta, age_predictions[,1], by=0)
df <- cbind(df, clust)
df <- mutate(df, prediction = y, time_point = TimePoint, actual=organism_age) %>% dplyr::select(-y, -TimePoint, -organism_age)
df <- df %>% mutate(actual = as.numeric(actual))

time_point_ordering <- c("Pre-Infection",
                         "0 Weeks",
                         "1 Week",
                         "2 Weeks",
                         "3 Weeks",
                         "4 Weeks",
                         "6 Months",
                         "1 Year")
replacer <- c("Pre_Infection",
               "0_Weeks",
               "1_Week",
               "2_Weeks",
               "3_Weeks",
               "4_Weeks",
               "6_Months",
               "1_Year")
names(replacer) <- time_point_ordering
df <- df %>% 
  mutate(time_point = factor(time_point, ordered=T,
                                        levels = time_point_ordering),
                    donor_id = factor(donor_id)) %>%
  mutate(time_point = recode(time_point, !!!replacer))%>%
  mutate(time_point = factor(time_point, ordered=F))

age_diff <- c(0, 0, 1/52, 2/52, 3/52, 4/52, 1/2, 1)
names(age_diff) <- c("Pre_Infection",
                     "0_Weeks",
                     "1_Week",
                     "2_Weeks",
                     "3_Weeks",
                     "4_Weeks",
                     "6_Months",
                     "1_Year")


df <- df  %>% mutate(actual = actual+age_diff[time_point])
df <- df  %>% mutate(residual = prediction-actual)

df <- df %>% filter(time_point=="Pre_Infection"|time_point=="1_Year")

## isolate data for Fig4F
cluster <- "0"
this_cluster <- df[df$clust==cluster,] %>% 
  dplyr::select(time_point, residual) %>%
  mutate(time_point = factor(time_point, ordered=F)) %>%
  droplevels()
tp_levels <- levels(this_cluster$time_point)
comps <-list(rep(tp_levels[1], length(tp_levels) -1),
                   tp_levels[2:length(tp_levels)])
print(cluster)
print(comps)
print(dim(this_cluster))
num_at_timepoint <- table(this_cluster$time_point)
for (i in names(num_at_timepoint)){
  #print(num_at_timepoint[i])
  if (num_at_timepoint[i] < 2){
    print(paste(i, "has", num_at_timepoint[i], "cells"))
    print(dim(this_cluster))
    print(levels(this_cluster$time_point))
    print(table(this_cluster$time_point))
    this_cluster <- this_cluster %>% filter(time_point!=i)
    this_cluster$time_point<-droplevels(this_cluster$time_point)
    print(dim(this_cluster))
    print(levels(this_cluster$time_point))
    print(table(this_cluster$time_point))
  }
}

stat.test <- this_cluster %>% 
  t_test(residual ~ time_point,
         ref.group = "Pre_Infection") %>%
  adjust_pvalue(method = "BH") %>%
  add_significance() %>%
  add_xy_position(x = "time_point")
print(stat.test)
  
# cluster_violin <- ggplot(data=this_cluster, aes(x=time_point, y=residual)) +
#   theme_minimal() + theme_bw()  +
# geom_jitter(aes(alpha = 0.5), position = position_dodge(width = 0.2), color = 'grey50') +
#   geom_violin(aes(alpha = 0.5, fill=time_point), width = 0.4) +
#   geom_boxplot(aes(alpha = 0, fill=time_point),width = 0.2, outlier.shape = NA, position = position_dodge(width = 0.1)) +
#   #facet_grid(cluster~., scales = 'free') +
#   scale_fill_manual(values = c('pink','red3'))+
#   stat_pvalue_manual(stat.test, label = "p.adj", size=3)
cluster_violin <- ggplot(data=this_cluster, aes(x=time_point, y=residual)) +
  geom_jitter(position = position_dodge(width = 0.2), color = 'black') +
  geom_violin(aes(fill=time_point)) +
  scale_fill_manual(values = c('lightGreen','forestGreen'))+
  theme_minimal() + theme_bw(base_size=6)  +
  geom_boxplot(width = 0.1) +
  stat_pvalue_manual(stat.test, label = "p.adj", size=3)
print("Figure 4E")
ggsave("4E_HIV_violin.png")
print(cluster_violin)
write.csv(this_cluster, "4E_HIV_violin_this_cluster.csv")
```

```{r Figure4G}
## code from 2.1 age_prediction, written by Humza Hemani.
library(tidyverse)
library(glmnet)
library(Seurat)
library(Matrix)

add_dummy_genes_to_matrix <- function(data_matrix, gene_list_original) {
  
  # identify missing genes:
  pbmc_genes <- rownames(data_matrix)
  genes_dummy <- gene_list_original[!(gene_list_original %in% pbmc_genes)]
  genes_absent <- pbmc_genes[!(pbmc_genes %in% gene_list_original)]

  print(paste("genes_dummy", length(genes_dummy)))
  print(paste("genes_absent",length(genes_absent)))
  
  # identify all possible names of genes that are 'missing'
  library(org.Hs.eg.db)
  convert <- AnnotationDbi::select(org.Hs.eg.db, keys = genes_absent, columns="SYMBOL",keytype="ALIAS")
  
  # extract the names of the genes that match
  converted_genes <- convert[convert$SYMBOL %in% genes_dummy,]
  colnames(converted_genes)[1] <- 'gene'
  converted_genes <- converted_genes %>% group_by(SYMBOL) %>% sample_n(1)
  converted_genes <- converted_genes[converted_genes$gene %in% pbmc_genes,]
  converted_genes <- converted_genes[converted_genes$SYMBOL %in% gene_list_original,]
  pbmc_df <- tibble(gene=pbmc_genes) %>% left_join(converted_genes, by = 'gene')
  pbmc_df <- pbmc_df %>% group_by(gene) %>% sample_n(1) %>% ungroup()
  pbmc_df2 <- tibble(gene=pbmc_genes) %>% left_join(pbmc_df, by = 'gene')
  pbmc_df <- pbmc_df2 %>% mutate(gene = ifelse(is.na(SYMBOL), gene, SYMBOL))
  
  newrownames <- pbmc_df$gene
  genes_dummy <- gene_list_original[!(gene_list_original %in% newrownames)]
  
  
  empty_mat <- matrix(rep(0, length(genes_dummy) * ncol(data_matrix)), ncol = ncol(data_matrix))
  rownames(empty_mat) <- genes_dummy
  
  
  rownames(data_matrix) <- newrownames
  test_data <- rbind(data_matrix, empty_mat)
  test_data <- test_data[gene_list_original,]
  
  return(test_data)
  
}

seu_obj <- readRDS("./1.3_only_CD8_reclustered.Rds")

all.dat <- seu_obj[['RNA']]@data
all.dat <- all.dat[(rownames(all.dat) != "Sex"),]
new_sex_row <- plyr::mapvalues(x = seu_obj$Sex,
                               from = c('M','F'),
                               to = c(1,0))# 0 for female, 1 for male.
new_sex_row <- as.numeric(levels(new_sex_row))[new_sex_row]
all.dat <- rbind(all.dat, Sex=new_sex_row) 

######
enet.model <- readRDS('./3.5.5_MEEN_glmnet_integrated_OneGrp_logNorm_parallel_geneNameFix.Rds')
fit.enet <- enet.model$fit.enet

model.features <- fit.enet[["glmnet.fit"]][["beta"]]@Dimnames[[1]]
all.dat <- add_dummy_genes_to_matrix(all.dat, model.features)

print("starting actual prediciton")
prediction <- predict(fit.enet, newx = t(all.dat), s="lambda.min")
saveRDS(prediction, file.path("age_predictions_car_t.Rds"))

## code from 2.2
library(rstatix)
library(ggpubr)

seu_metadata <- readRDS("./1.3_metadata.Rds")
age_prediction <- data.frame(prediction=readRDS("./age_predictions_car_t.Rds")) %>%
  mutate(prediction=`lambda.min`)

# combine the relevate seurat metadata and the age predictions
df <- seu_metadata %>%
  dplyr::select(seurat_clusters,
         Patient, 
         Group, # Group variable refers to the text-based time point e.g. "IP", "Late"
         Disease,
         TimePoint, # TimePoint variable refers to the exact day
         Sex) %>%
        cbind(age_prediction)
df <- df %>% mutate(day = plyr::mapvalues(df$TimePoint,
                                          from = c("d102","d112","d12","d21","d28","d29","d38","d83","d89","IP"),
                                          to = c(102, 112, 12,21,28,29,38,83,89,0))) %>%
  mutate(actual =
               plyr::mapvalues(
                 x = df$Patient,
                 from = c('NHL-9','NHL-10','CLL-1','CLL-2'),
                 to = c(64, 51, 61, 53)
               ) # actual age of each patient
  )
# adjust the actual age of the patients to account for time since diagnosis
df <- df %>% mutate(actual = as.numeric(levels(actual))[actual] + as.numeric(levels(day))[day]/365)

# calculate the difference between the predicted age and actual age
df <- df %>% mutate(residual = prediction - actual)

# Remove Contraction timepoint
# Only looking at IP, Expansion Peak, and Late.
df2 <- df %>% 
  filter(Group!="Contraction") %>%
  mutate(day=droplevels(day), Group=droplevels(Group))

# use the average residual of each person's cells at day 0 as a reference
# subtract that reference from each cell's residual in the dataset
# so we can clearly see how the CAR T-cells age after infusion
ref_map <- numeric(4)
names(ref_map) <- unique(df2$Patient)
for(patient in unique(df2$Patient)){
  wanted_values <- df2$residual[df2$Patient==patient & df2$Group=="IP"]
  print(patient)
  print(mean(wanted_values))
  ref_map[patient] <- mean(wanted_values)
  summary(wanted_values)
  summary(wanted_values-mean(wanted_values))
}
df2 <- df2%>%
  mutate(ref_age=plyr::mapvalues(Patient,
                                 from=names(ref_map),
                                 to=ref_map)) %>%
  mutate(ref_age = as.numeric(levels(ref_age))[ref_age]) %>%
  mutate(adjusted_residual=residual-ref_age)

# get the mean adjusted residual for each patient at each time point
df3 <- df2 %>% group_by(Patient,Group) %>% 
  dplyr::select(actual, prediction, residual, ref_age, adjusted_residual) %>%
  summarise_all(mean) %>% 
  ungroup() %>%
  mutate(Group=factor(Group, levels=c("IP", "ExpansionPeak", "Late")))

# get p values
stat.test <- df3 %>% 
  t_test(residual ~ Group) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance() %>%
  add_xy_position(x = "Group", y.trans=function(x){x+18})
print(stat.test)
# write.csv(stat.test, "overall_stat_test.csv")

# the rest of the script does plotting.
overall_combined <- ggplot(df3, 
                                  aes(x=Group, 
                                      y=adjusted_residual, 
                                      group=Group)) +
  geom_jitter(aes(alpha = 1.0),  position = position_dodge(width = 0.2), color = 'black') +
  #geom_violin(data = filter(df3, Group == "ExpansionPeak" | Group == "Late"),
              #aes(fill=Group)) +
  theme_minimal() + theme_bw() +
  geom_boxplot(data = filter(df3, Group == "ExpansionPeak" | Group == "Late"),
               width = 0.5, aes(fill=Group)) +
  stat_pvalue_manual(data=stat.test, label="p.adj", inherit.aes=F, size=3)

overall_with_mean_line <- overall_combined +
  stat_summary(aes(group=1),color='black',fun=mean, geom="line")
print("Figure 4G")
ggsave("fig_4g.png")
print(overall_with_mean_line)
write.csv(df3, "fig4g_df3.csv")
```

